js/src/Components/Widgets/CkeditorWidget.js

```diff
@@ -68,51 +48,65 @@ export class CkeditorWidget extends Component {
     }
   }
 
-  onFieldChange(value, format, summary) {
+  addListeners() {
+    // UnlockSnapshot is an event that is triggered when a button in the toolbar
+    // or media is added to the wysiwyg. Key is the only event triggered when
+    // the ckeditor is in "view source" mode.
+    CKEDITOR.instances[`${this.props.fieldId}-text-area`].on('unlockSnapshot', this.onEditorChange);
+    CKEDITOR.instances[`${this.props.fieldId}-text-area`].on("key", this.onEditorChange);
+  }
+
+  updateState(key, value) {
+    const newState = {...this.state};
+    newState[key] = value;
+    this.setState(newState);
+
+    // If the ckeditor area or the summary is empty, clear the field value.
+    if (newState.value.length === 0 && newState.summary.length === 0) {
+      this.props.onFieldChange([]);
+      return;
+    }
+
+    // Construct the new field value to bubble up to the widget manager.
     const newValue = [{
-      value: value,
-      format: format
+      value: newState.value,
+      format: newState.format
     }];
+
     if (this.props.settings.summary) {
-      newValue[0].summary = summary;
+      newValue[0].summary = newState.summary;
     }
     this.props.onFieldChange(newValue);
   }
 
-  onEditorChange(element, newValue) {
-    this.onFieldChange(newValue, this.state.format, this.state.summary);
-    this.setState({value: newValue});
-  }
+  onEditorChange(snapshot) {
+    clearTimeout(this.editorTimeout);
 
-  onTextAreaChange(event) {
-    this.onEditorChange(event, event.target.value);
-  }
+    // Make a new timeout set to go off in 200ms.
+    // This reduces the unnecessary calls since there are multiple event
+    // listeners on the ckeditor area.
+    this.editorTimeout = setTimeout(() => {
```

Comment:
I'm not positive on this but this is probably a good opportunity to turn this state update into an asynchronous task using a promise as you can start and cancel a promise if something else comes up before the call is complete.